bip-v1-packages.mediawiki

```diff
@@ -0,0 +1,314 @@
+
+# Upon receipt of a "pckginfo1" message, the node should decide if it wants to validate the package, request transaction data if necessary, etc.
+
+# Upon receipt of a malformed "pckginfo1" message or package that does not abide by the max_count, max_weight, or other rules specified by the version agreed upon in the initial negotiation, the sender should be disconnected.  If a node receives a "pckginfo1" message for which the "pckg_fee" or "pckg_weight" do not reflect the true total fees and weight, respectively, or the transactions in the package, the message is malformed.
+
+# A node MUST NOT send a "pckginfo1" message that has not been requested by the recipient. Upon receipt of an unsolicited "pckginfo1", a node should disconnect the sender.
+
+# A "pckginfo1" message should only be sent if both peers agreed to send version 1 packages in the version handshake. If a "pckginfo1" message is received from a peer with which package relay was not negotiated, the sender should be disconnected.
+
+====MSG_PCKG1====
+
+# A new inv type (MSG_PCKG1 == 0x6) is added, for use in inv messages and getdata requests pertaining to version 1 packages.
+
+# As an inv type, it indicates that both transaction data and version 1 package information are available for the transaction. The transaction is referenced by its wtxid. As a getdata request type, it indicates that the sender wants package information for the transaction.
+
+# Upon receipt of a "getdata" request for "MSG_PCKG1", the node should respond with the version 1 package corresponding to the requested transaction and its current chain tip, or with NOTFOUND. The node should not assume that the sender is requesting the transaction data as well.
+
+===Child With Parent Packages===
+
+Version 1 package relay supports child-with-unconfirmed-parents packages. All v1 packages
+sent between nodes MUST abide by the rules below, otherwise the package is malformed and the sender
+should be disconnected.
+
+A version 1 or '''child-with-unconfirmed-parents''' package can be defined for any transaction that spends
+unconfirmed inputs. The child can be thought of as the "representative" of the package. This package
+can be uniquely identified by the transaction's wtxid and the current chain tip block hash.
+
+A '''child-with-unconfirmed-parents''' package MUST be:
+
+# '''Sorted topologically.''' For every transaction t in the package, if any of t's parents are present in the package, the parent must appear somewhere in the list before t. In other words, the transactions must be sorted in ascending order of number of ancestors present in the package.
+
+# '''Only 1 child with unconfirmed parents.''' The package must consist of one transaction and its unconfirmed parents. There must not be any other transactions in the package. Other dependency relationships may exist within the package (e.g. one parent may spend the output of another parent) provided that topological order is respected.
```

Comment:
One thing I overlooked when I first read this over -- by "unconfirmed parents" you mean direct parents, ie transactions which contain an outpoint being spent in the child transaction, is that right?  (I was trying to figure out exactly what you meant by the 2-generation limit further down, and then I realized I might have misunderstood the language here.)

I'd suggest defining this language more explicitly somehow, since I think "unconfirmed parent" can read a lot like "unconfirmed ancestor" (and at any rate I don't think these words have meanings well defined across the developer community).

Also, if a `GETDATA(PCKG1, wtxid)` comes in for a transaction that has unconfirmed ancestors outside the set of unconfirmed direct parents, how is software supposed to respond?

Reply:
> by "unconfirmed parents" you mean direct parents, ie transactions which contain an outpoint being spent in the child transaction, is that right?

Yes, exactly.

> I'd suggest defining this language more explicitly somehow, since I think "unconfirmed parent" can read a lot like "unconfirmed ancestor" (and at any rate I don't think these words have meanings well defined across the developer community).

Ah, noted. How about something like this:
"Given any two transactions Tx0 and Tx1 where Tx1 spends an output of Tx0, Tx0 is a parent of Tx1 and Tx1 is a child of Tx0. A transaction's ancestors include, recursively, its parents, the parents of its parents, etc. A transaction's descendants include, recursively, its children, the children of its children, etc."

> Also, if a GETDATA(PCKG1, wtxid) comes in for a transaction that has unconfirmed ancestors outside the set of unconfirmed direct parents, how is software supposed to respond?

It should send only the direct parents, and no other ancestors.